src/main/java/build/buildfarm/metrics/cas/CASAccessMetricsRecorder.java

```diff
@@ -0,0 +1,242 @@
+
+  final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+  int intervalNumber = 0; // Assists in unit test.
+
+  public CASAccessMetricsRecorder(
+      ScheduledExecutorService scheduler,
+      Backplane backplane,
+      Duration casEntryReadCountWindow,
+      Duration casEntryReadCountUpdateInterval) {
+    Preconditions.checkArgument(
+        casEntryReadCountUpdateInterval.toMillis() > 0,
+        "CASEntryReadCountUpdateInterval must be greater than 0");
+    Preconditions.checkArgument(
+        casEntryReadCountWindow.toMillis() % casEntryReadCountUpdateInterval.toMillis() == 0,
+        "CASEntryReadCountWindow must be divisible by CASEntryReadCountUpdateInterval");
+    this.scheduler = scheduler;
+    this.backplane = backplane;
+    this.casEntryReadCountWindow = casEntryReadCountWindow;
+    this.casEntryReadCountUpdateInterval = casEntryReadCountUpdateInterval;
+    this.readIntervalCountQueue = new LinkedList<>();
+    this.queueLength = getQueueLength();
+  }
+
+  /**
+   * Increments the read count by 1 for the given digest.
+   *
+   * @param digest Digest to which the read count is recorded.
+   */
+  public void recordRead(Digest digest) {
+    if (!running) {
+      throw new IllegalStateException("Metrics Recorder is not running");
+    }
+    runWithLock(
+        lock.readLock(),
+        () ->
+            currentIntervalReadCount
+                .computeIfAbsent(digest, d -> new AtomicInteger(0))
+                .incrementAndGet());
+  }
+
+  /**
+   * Records the write operation for a digest if it is new. If a read is recorded before for a newly
+   * written digest, this method will have no effect.
+   *
+   * @param digest Digest for which the write operation is recorded.
+   */
+  public void recordWrite(Digest digest) {
+    if (!running) {
+      throw new IllegalStateException("Metrics Recorder is not running");
+    }
+    runWithLock(
+        lock.readLock(), () -> currentIntervalReadCount.putIfAbsent(digest, new AtomicInteger(0)));
+  }
+
+  /**
+   * Transitions the metrics recorder to the running state and sets up periodic read updates. Once
+   * the metrics recorder is in the running state, calling this method will have no effect.
+   */
+  public synchronized void start() {
+    if (running) {
+      return;
+    }
+    currentIntervalReadCount = new ConcurrentHashMap<>();
+    long initialDelay = getInitialDelayInMillis();
+    scheduler.scheduleAtFixedRate(
+        this::updateReadCount,
+        initialDelay,
+        casEntryReadCountUpdateInterval.toMillis(),
+        TimeUnit.MILLISECONDS);
+    running = true;
+  }
+
+  /** Stops the scheduler responsible to update read counts periodically. */
+  public void stop() throws InterruptedException {
+    running = false;
+    scheduler.shutdown();
+    if (!scheduler.awaitTermination(10, TimeUnit.SECONDS)) {
+      log.severe("Could not shut down cas access metrics recorder service");
+    }
+    scheduler.shutdownNow();
```

Comment:
This class should not presume ownership of the scheduler it is passed - stop the recorder via this method, and stop the scheduler independently, owned (and maintained) by the caller

Reply:
Moved logic to stop the scheduler in Worker class